examples/rifle_sxiv.sh

```diff
@@ -11,38 +11,38 @@
 #   mime ^image, has sxiv, X, flag f = path/to/this/script -- "$@"
 #
 # Implementation notes: this script is quite slow because of POSIX limitations
-# and portability concerns. First calling the shell function 'abspath' is
+# and portability concerns. First, using case statement to get absolute path is
 # quicker than calling 'realpath' because it would fork a whole process, which
 # is slow. Second, we need to append a file list to sxiv, which can only be done
 # properly in two ways: arrays (which are not POSIX) or \0 sperated
 # strings. Unfortunately, assigning \0 to a variable is not POSIX either (will
 # not work in dash and others), so we cannot store the result of listfiles to a
 # variable.
 
-if [ $# -eq 0 ]; then
-    echo "Usage: ${0##*/} PICTURES"
-    exit
-fi
-
-[ "$1" = '--' ] && shift
+listfiles () {
+    find -L "///${target%/*}" -maxdepth 1 -type f -iregex \
+      '.*\.\(jpe?g\|png\|gif\|webp\|tiff\|bmp\)$' -print0 | sort -z
+}
 
-abspath () {
-    case "$1" in
-        /*) printf "%s\n" "$1";;
-        *)  printf "%s\n" "$PWD/$1";;
+is_img () {
+    case "${1##*.}" in
+        "jpg"|"jpeg"|"png"|"gif"|"webp"|"tiff"|"bmp") return 0 ;;
+        *) return 1 ;;
     esac
 }
 
-listfiles () {
-    find -L "$(dirname "$target")" -maxdepth 1 -type f -iregex \
-      '.*\(jpe?g\|bmp\|png\|gif\)$' -print0 | sort -z
-}
+[ "$1" = '--' ] && shift
+case "$1" in
+    "") echo "Usage: ${0##*/} PICTURES" >/dev/stderr && exit ;;
+    /*) target="$1" ;;
+    "~"/*) target="$HOME/${1#"~"/}" ;;
+    *)  target="$PWD/$1" ;;
+esac
 
-target="$(abspath "$1")"
-count="$(listfiles | grep -m 1 -ZznF "$target" | cut -d: -f1)"
+is_img "$target" && count="$(listfiles | grep -m 1 -ZznF "$target")"
```

Comment:
Likewise, we could add the much faster ripgrep and relegate grep to being a fallback:

```suggestion
is_img "$target" && count="$(listfiles | rg -Fnm1 --null-data "$target" || grep -Fnzm1 "$target")"
```

I believe `-Z` is redundant in the presence of `-z`. BSD grep doesn't support the `-z/--null-data` option unfortunately so this wouldn't work with that grep.

Reply:
The reason we needed `-z` was so we can feed it to `sxiv` sanely. We can add an argument to `listfiles` to print or not print null so grep wouldn't need `-z`.